### PR TITLE
Refactor: refactor stub-components.js

### PR DESCRIPTION
**Pull request by lusarz**
Pull request contains changes like
- separate stub options validation from creation of stubs
- overall cleaning

### CODE REVIEW

File: packages/shared/stub-components.js [hunk 1]
```diff
@@ -40,147 +25,72 @@ function getCoreProperties (component: Component): Object {
...
+      return object
+    }, {})
   }
+  return stubOptions
+}
 
-  if (component.extendOptions && component.extendOptions.components) {
-    stubComponents(component.extendOptions.components, stubbedComponents)
-  }
+export function createComponentStubs (components: Object = {}, stubOptions: Object): Object {
```
Comment (lusarz) [inline packages/shared/stub-components.js:70]: Personally I think public methods should be moved to top of file. Didn't do it to satisfy `git blame` as much as possible

File: packages/shared/stub-components.js [hunk 2]
```diff
@@ -40,147 +25,72 @@ function getCoreProperties (component: Component): Object {
...
+    .filter(name => options[name] !== false)
+    .reduce((stubs, name) => {
+      stubs[name] = createStub(components[name], options[name])
 
-    components[c] = createBlankStub(instance.options.components[c])
-    delete instance.options.components[c]._Ctor // eslint-disable-line no-param-reassign
-    delete components[c]._Ctor // eslint-disable-line no-param-reassign
-  })
-  return components
+      if (Vue.config.ignoredElements) {
```
Comment (lusarz) [inline packages/shared/stub-components.js]: Do we need it ? I copied it from original source.

Comment (eddyerburgh) [reply packages/shared/stub-components.js]: I don't think we do. you can remove it :)

File: packages/shared/stub-components.js (hunk 1 above)
Comment (eddyerburgh) [reply packages/shared/stub-components.js:70]: I don't think we should move the position of functions. Most files in this project have public functions at the bottom of the files

File: packages/shared/stub-components-validate.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  } else {
+    if (containsInvalidOptions(stubOptions)) {
+      throwError('options.stub values must be passed a string or component')
+    }
+
+    if (necessaryCompileToFunctionsMissed(stubOptions)) {
+      throwError('vueTemplateCompiler is undefined, you must pass components explicitly if vue-template-compiler is undefined')
+    }
+
+    if (containsTemplateWithCircularReference(stubOptions, components)) {
```
Comment (eddyerburgh) [inline packages/shared/stub-components-validate.js]: This is actually causing bugs for some users. The check is not strict enough, so I've removed the check in the source code for the moment. Could you either remove this code, or improve the code to handle edge cases like: `UnreadState: '<div data-stub-unread-state v-bind="$props"><slot/></div>'`

File: packages/shared/stub-components.js (hunk 1 above)
Comment (lusarz) [reply packages/shared/stub-components.js:70]: Ok, let it stay like that

File: packages/shared/stub-components.js (hunk 2 above)
Comment (lusarz) [reply packages/shared/stub-components.js]: Removed :)

File: packages/shared/stub-components-validate.js (hunk 3 above)
Comment (lusarz) [reply packages/shared/stub-components-validate.js]: Removed :)

**Review by eddyerburgh — APPROVED** [review]

**Comment by eddyerburgh** [comment]
Thanks :)

**Pull request merged**